### PR TITLE
[TRANSFORMATIONS] Introduce a new method of testing SDPAToPA transformation

### PR DESCRIPTION
**Pull request by CuriousPanCake**
[TRANSFORMATIONS] Introduce a new method of testing SDPAToPA transformation

Introduce a new method of testing the SDPAToPA transformation by not only checking if PagedAttentionExtension nodes appeared in a graph, but also check if the changes in number of nodes involved in the transformation aligned with the reference numbers for each model.

Add a script for fast generation of reference values.

Signed-off-by: Andrii Staikov <andrii.staikov@intel.com>

- Tickets:
	* CVS-152290

### CODE REVIEW

File: tests/model_hub_tests/transformation_tests/generate_model_diffs.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+'''
```
Comment (dorloff) [inline tests/model_hub_tests/transformation_tests/generate_model_diffs.py:1]: We have three terms: "model_diffs" in this script name, "model_diff_ref" in the generated file names, and "pa_reference" in the variable name for the same thing. I would choose one term.

File: tests/model_hub_tests/transformation_tests/generate_model_diffs.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+import os
+import sys
+from pathlib import Path
+import models_hub_common.utils as utils
+from openvino._offline_transformations import paged_attention_transformation
+from openvino._pyopenvino.op import _PagedAttentionExtension, Parameter, Result
+from optimum.intel import OVModelForCausalLM
+
+explored_nodes = ("ScaledDotProductAttention", "PagedAttentionExtension", "Parameter", "ReadValue", "Assign")
```
Comment (dorloff) [inline tests/model_hub_tests/transformation_tests/generate_model_diffs.py:44]: Maybe "nodes_to_compare"?

Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/generate_model_diffs.py:44]: Fixed

File: tests/model_hub_tests/transformation_tests/generate_model_diffs.py (hunk 1 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/generate_model_diffs.py:1]: Fixed. Aligned all the names

File: tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py [hunk 3]
```diff
@@ -0,0 +1,595 @@
+nodes_to_compare = ("ScaledDotProductAttention", "PagedAttentionExtension", "Parameter", "ReadValue", "Assign")
```
Comment (rkazants) [inline tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py:1]: copyright?

File: tests/model_hub_tests/transformation_tests/generate_ref_diffs.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
+'''
```
Comment (rkazants) [inline tests/model_hub_tests/transformation_tests/generate_ref_diffs.py:1]: copyright?

File: tests/model_hub_tests/transformation_tests/generate_ref_diffs.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+Use this script if you need to regenerate reference diffs for each model
+to test SDPAToPA transformation.
+
+The script will produce sdpa2pa_ref_diff.txt (or sdpa2pa_ref_diff_ce.txt
+if using cache-eviction) containing a map in the
+following format with nodes number changes for each model:
+
```
Comment (rkazants) [inline tests/model_hub_tests/transformation_tests/generate_ref_diffs.py:8]: how is this approach for testing is scalable to other platforms?

File: tests/model_hub_tests/transformation_tests/generate_ref_diffs.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	"hf-internal-testing/tiny-random-CohereForCausalLM" : {
+		"PagedAttentionExtension" : 2,
+		"ScaledDotProductAttention" : -2,
+		"Parameter" : 7,
+		"ReadValue" : -4,
+		"Assign" : -4,
+	},
+    .
+    .
+    .
```
Comment (rkazants) [inline tests/model_hub_tests/transformation_tests/generate_ref_diffs.py:26]: ?

File: tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py (hunk 3 above)
Comment (rkazants) [inline tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py:1]: why this is of `.py` extension. Is it python file?

File: tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+
+ref_diff_map = {
+	"hf-internal-testing/tiny-random-LlamaForCausalLM" : {
+		"PagedAttentionExtension" : 2,
+		"ScaledDotProductAttention" : -2,
+		"Parameter" : 7,
+		"ReadValue" : -4,
+		"Assign" : -4,
```
Comment (rkazants) [inline tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py:9]: can you please give comments where are these values coming from?

File: tests/model_hub_tests/transformation_tests/generate_ref_diffs.py (hunk 6 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/generate_ref_diffs.py:26]: This is like "...", but vertical :)
Just to show that there're other models present in the resulting map.

File: tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py (hunk 3 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py:1]: Yes, this is a Python file that is being included in test_pa_transformation.py later on.

File: tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py (hunk 7 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py:9]: These values were generated using generate_model_diffs.py. If someone's making a change that would introduce some change in the number of nodes, the values can be regenerated.

File: tests/model_hub_tests/transformation_tests/models/hf-tiny-random-models-precommit [hunk 8]
```diff
@@ -41,6 +41,4 @@ facebook/opt-125m,https://huggingface.co/facebook/opt-125m
 facebook/opt-350m,https://huggingface.co/facebook/opt-350m
 katuni4ka/tiny-random-chatglm2,https://huggingface.co/katuni4ka/tiny-random-chatglm2
 katuni4ka/tiny-random-glm4,https://huggingface.co/katuni4ka/tiny-random-glm4
-hf-internal-testing/tiny-random-BioGptForCausalLM,https://huggingface.co/hf-internal-testing/tiny-random-BioGptForCausalLM,xfail,No ScaledDotProductAttention operation observed in the graph CVS-145820
-hf-internal-testing/tiny-random-XGLMForCausalLM,https://huggingface.co/hf-tiny-model-private/tiny-random-XGLMForCausalLM,xfail,No ScaledDotProductAttention operation observed in the graph CVS-145820
 katuni4ka/tiny-random-orion,https://huggingface.co/katuni4ka/tiny-random-orion,xfail,No ScaledDotProductAttention operation observed in the graph CVS-145820
```
[the remark this answers is not in the document]
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/models/hf-tiny-random-models-precommit]: Done

File: tests/model_hub_tests/transformation_tests/generate_ref_diffs.py (hunk 4 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/generate_ref_diffs.py:1]: Fixed.

File: tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py (hunk 3 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py:1]: Fixed.

File: tests/model_hub_tests/transformation_tests/test_pa_transformation.py [hunk 9]
```diff
@@ -14,15 +15,28 @@
...
+    # And check if the numbers correspond to the reference ones
+    resulting_map = {}
+    for op in set(after_map.keys()) | set(before_map.keys()):
+        resulting_map[op] = after_map.get(op, 0) - before_map.get(op, 0)
+
+    use_cache_eviction = use_block_indices_inputs and use_score_outputs
+    reference_map = ref_diff_map_ce[model_id] if use_cache_eviction else ref_diff_map[model_id]
 
-    assert pa_counter > 0, f"The model '{model_id}' has no _PagedAttentionExtension present."
+    assert reference_map == resulting_map
```
Comment (itikhono) [inline tests/model_hub_tests/transformation_tests/test_pa_transformation.py:39]: it's better to print a more verbose message
e.g. 
Unexpected changes were found after applying the SDPAtoPagedAttention transformation.
Parameters Expected: +3 Actual: +2
etc

File: tests/model_hub_tests/transformation_tests/test_pa_transformation.py [hunk 10]
```diff
@@ -45,7 +59,7 @@ def run_pa(tmp_path, model_id, model_link, use_block_indices_inputs, use_score_o
                 if re.search(block_indices_pattern, name):
                     block_indices_counter += 1
 
-        assert(block_indices_counter == pa_counter)
+        assert(block_indices_counter == resulting_map["PagedAttentionExtension"])
```
Comment (itikhono) [inline tests/model_hub_tests/transformation_tests/test_pa_transformation.py:62]: print more verbose messages

The number of BlockIndices is expected to be equal to the number of PagedAttentionExtension ops 
Block indices cnt: block_indices_counter
 Paged Attention cnt: resulting_map["PagedAttentionExtension"]

File: tests/model_hub_tests/transformation_tests/test_pa_transformation.py [hunk 11]
```diff
@@ -57,7 +71,7 @@ def run_pa(tmp_path, model_id, model_link, use_block_indices_inputs, use_score_o
                 if re.search(score_pattern, name):
                     score_outputs_counter += 1
 
-        assert(score_outputs_counter == pa_counter)
+        assert(score_outputs_counter == resulting_map["PagedAttentionExtension"])
```
Comment (itikhono) [inline tests/model_hub_tests/transformation_tests/test_pa_transformation.py:74]: print more verbose messages

File: tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py [hunk 12]
```diff
@@ -0,0 +1,612 @@
...
+	"katuni4ka/tiny-random-glm4" : {
+		"PagedAttentionExtension" : 6,
+		"ScaledDotProductAttention" : -6,
+		"Parameter" : 15,
+		"ReadValue" : -12,
+		"Assign" : -12,
+	},
+}
+
+ref_diff_map_ce = {
```
Comment (itikhono) [inline tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py:310]: "ce" ?

Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py:310]: cache-eviction

File: tests/model_hub_tests/transformation_tests/test_pa_transformation.py (hunk 9 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/test_pa_transformation.py:39]: Here python would print a verbose message by itself, no worries here

File: tests/model_hub_tests/transformation_tests/test_pa_transformation.py (hunk 10 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/test_pa_transformation.py:62]: Yeah, I need  to check in this place and add if needed

File: tests/model_hub_tests/transformation_tests/test_pa_transformation.py (hunk 11 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/test_pa_transformation.py:74]: Added a more verbose assertion message

image

File: tests/model_hub_tests/transformation_tests/test_pa_transformation.py (hunk 10 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/test_pa_transformation.py:62]: Added a more verbose assertion message:

image

File: tests/model_hub_tests/transformation_tests/test_pa_transformation.py (hunk 9 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/test_pa_transformation.py:39]: Already a pretty verbose message:

image

File: tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py (hunk 12 above)
Comment (CuriousPanCake) [reply tests/model_hub_tests/transformation_tests/sdpa2pa_ref_diff.py]: Fixed. aded _cache_eviction

**Review by itikhono — APPROVED** [review]

**Review by dorloff — APPROVED** [review]

**Pull request merged**